src/EditorFeatures/Core/Implementation/FindReferences/AbstractFindReferencesService.cs

```diff
@@ -51,36 +59,127 @@ protected AbstractFindReferencesService(IEnumerable<IReferencedSymbolsPresenter>
             return null;
         }
 
-        public async Task<IEnumerable<INavigableItem>> FindReferencesAsync(Document document, int position, IWaitContext waitContext)
+        /// <summary>
+        /// Finds references using the externally defined <see cref="IFindReferencesResultProvider"/>s.
+        /// </summary>
+        private async Task AddExternalReferencesAsync(Document document, int position, ArrayBuilder<INavigableItem> builder, CancellationToken cancellationToken)
+        {
+            if (_externalReferencesProviders.Any())
+            {
+                foreach (var provider in _externalReferencesProviders)
```

Comment:
Given that these are MEF based providers, I am not sure if running in parallel can cause any deadlocks and unforeseen side effects. I followed the sequential approach used by CodeFixService to compute fixes - we should definitely think about this aspect when discussing this at the design meeting.